### PR TITLE
add discriminant to Statement

### PR DESCRIPTION
**Pull request by WalterBright**
Making it more like Expression and Type, which will enable getting rid of many virtual functions and Visitor implementations.

### CODE REVIEW

**Comment by JinShil** [comment]
> which will enable getting rid of many virtual functions and Visitor implementations.

What's wrong with virtual functions and visitors? Is this change for performance? If so, has the difference in performance been measured and is it significant?

**Comment by WalterBright** [comment]
> What's wrong with virtual functions and visitors?

Performance, complexity, and problems making things `const`. Virtual dispatch is quite a bit slower than testing a constant (not cache friendly for one thing), and the visitor pattern requires two virtual dispatches.

> Is this change for performance?

Partly. It's also for encapsulation and difficulty in adding attributes to the Visitor thing.

> If so, has the difference in performance been measured and is it significant?

It's significant enough that many efforts have been made to de-virtualize virtual function calls. Back when dmd first started using Visitor, it was measured and slowed dmd down by "only 1%" or something like that. I've seen that "only 1%" many times. 10 of those, and you've got 10%. Pretty soon, dmd is as slow as clang. It's rarely just one barnacle that makes a compiler slow, it's an encrusted hull that does it.

**Comment by andralex** [comment]
This approach seems to have a number of issues. I suggest holding off on pulling it. The previous design used classic OOP to distinguish types. Under sufficient justification devirtualization is sensible, but this code keeps both classes and the tag, essentially creating an unholy redundancy. Now it has the drawbacks of classes and those of tag dispatch, plus a few resulting from the redundancy.

**Comment by WalterBright** [comment]
I intended for subsequent PRs to make use of the tag. Besides, this approach is used for both Type and Expression throughout the code. When I'd remove the use of the Visitor code from them, the code became substantially smaller, with no dependencies on other externalities (like the Visitor implementation).

**Review by ibuclaw — APPROVED** [review]
Indeed, the desired target here is to turn free functions that use a nested visitor into free functions that use a switch table.

**Review by PetarKirov — APPROVED** [review]
I'm also strongly in favor of replacing the use of the visitor pattern for cases like this, where an explicit switch on a discriminant field brings benefits such as easier application of attributes like `const` and `pure`, in addition to the above mentioned potential perf improvements.

**Comment by UplinkCoder** [comment]
We should really at a discriminant to `RootObject` while we are at it.
th

**Review by UplinkCoder — APPROVED** [review]
Good direction to take!

File: src/dmd/astbase.d [hunk 1]
```diff
@@ -1880,13 +1880,58 @@ struct ASTBase
 
     }
 
+    enum STMT : ubyte
```
Comment (Basile-z) [inline src/dmd/astbase.d]: sadly this is redefined a second time. Remember the PR for STC a few weeks ago.

**Review by ghost — APPROVED** [review]
In addition to the positive points enumerated previously the tag are also somewhat useful with `printf` style debugging.

**Comment by wilzbach** [comment]
> It's rarely just one barnacle that makes a compiler slow, it's an encrusted hull that does it.

Fair enough, but there's still this very low hanging fruit of compiling DMD with LDC (which would result in DMD being twice as fast) ;-)

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**